Eureka Client cannot make HTTPS requests to servers that use multiple certificates via SNI

EurekaJerseyClientImpl uses deprecated, pre HTTPClient 4.3 APIs to create the SSLSocketFactory used in MonitoredConnectionManager [[1]](https://github.com/Netflix/eureka/blob/v1.6.0/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/EurekaJerseyClientImpl.java#L222). Support for https://en.wikipedia.org/wiki/Server_Name_Indication was introduced in Apache HttpClient 4.3.2 [[2]](https://hc.apache.org/news.html). When using the deprecated APIs, requests to servers that are protected with multiple certificates, which is typical of environments with many applications behind a single load balancer, will fail.

This can be demonstrated by setting the serviceUrl to `https://api.google.com/`. If the SSL handshake doesn't do SNI properly, the client will get the certificate for `www.google.com`, and hostname verification will fail with this exception. If SNI succeeds we won't see this exception (but we'll see another Eureka specific exception obviously).

```
2016-11-15 17:12:33.841 ERROR 65408 --- [nfoReplicator-0] c.n.d.s.t.d.RedirectingEurekaHttpClient  : Request execution error

com.sun.jersey.api.client.ClientHandlerException: javax.net.ssl.SSLException: Certificate for <api.google.com> doesn't match any of the subject alternative names: [www.google.com]
	at com.sun.jersey.client.apache4.ApacheHttpClient4Handler.handle(ApacheHttpClient4Handler.java:187) ~[jersey-apache-client4-1.19.1.jar:1.19.1]
	at com.sun.jersey.api.client.filter.GZIPContentEncodingFilter.handle(GZIPContentEncodingFilter.java:123) ~[jersey-client-1.19.1.jar:1.19.1]
	at com.netflix.discovery.EurekaIdentityHeaderFilter.handle(EurekaIdentityHeaderFilter.java:27) ~[eureka-client-1.4.11.jar:1.4.11]

...

Caused by: javax.net.ssl.SSLException: Certificate for <api.google.com> doesn't match any of the subject alternative names: [www.google.com]
	at org.apache.http.conn.ssl.AbstractVerifier.verify(AbstractVerifier.java:165) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.BrowserCompatHostnameVerifier.verify(BrowserCompatHostnameVerifier.java:61) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.AbstractVerifier.verify(AbstractVerifier.java:141) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.AbstractVerifier.verify(AbstractVerifier.java:114) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.SSLSocketFactory.verifyHostname(SSLSocketFactory.java:580) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:554) ~[httpclient-4.5.2.jar:4.5.2]
	at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:412) ~[httpclient-4.5.2.jar:4.5.2]
	at 
...
``` 

[1] https://github.com/Netflix/eureka/blob/v1.6.0/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/EurekaJerseyClientImpl.java#L222
[2] https://hc.apache.org/news.html

--- a/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/EurekaJerseyClientImpl.java
+++ b/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/EurekaJerseyClientImpl.java
@@ -17,9 +17,13 @@ import com.sun.jersey.client.apache4.ApacheHttpClient4;
 import com.sun.jersey.client.apache4.config.ApacheHttpClient4Config;
 import com.sun.jersey.client.apache4.config.DefaultApacheHttpClient4Config;
 import org.apache.http.client.params.ClientPNames;
+import org.apache.http.conn.scheme.PlainSocketFactory;
 import org.apache.http.conn.scheme.Scheme;
 import org.apache.http.conn.scheme.SchemeRegistry;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
 import org.apache.http.conn.ssl.SSLSocketFactory;
+import org.apache.http.conn.ssl.X509HostnameVerifier;
+import org.apache.http.impl.conn.SchemeRegistryFactory;
 import org.apache.http.params.CoreProtocolPNames;
 import org.apache.http.params.HttpConnectionParams;
 import org.apache.http.params.HttpParams;
@@ -180,7 +184,7 @@ public class EurekaJerseyClientImpl implements EurekaJerseyClient {
                 } else if (trustStoreFileName != null) {
                     cm = createCustomSslCM();
                 } else {
-                    cm = new MonitoredConnectionManager(clientName);
+                    cm = createDefaultSslCM();
                 }
 
                 if (proxyHost != null) {
@@ -219,7 +223,8 @@ public class EurekaJerseyClientImpl implements EurekaJerseyClient {
 
             private MonitoredConnectionManager createSystemSslCM() {
                 MonitoredConnectionManager cm;
-                SSLSocketFactory sslSocketFactory = SSLSocketFactory.getSystemSocketFactory();
+                SSLConnectionSocketFactory systemSocketFactory = SSLConnectionSocketFactory.getSystemSocketFactory();
+                SSLSocketFactory sslSocketFactory = new SSLSocketFactoryAdapter(systemSocketFactory);
                 SchemeRegistry sslSchemeRegistry = new SchemeRegistry();
                 sslSchemeRegistry.register(new Scheme(PROTOCOL, HTTPS_PORT, sslSocketFactory));
                 cm = new MonitoredConnectionManager(clientName, sslSchemeRegistry);
@@ -241,8 +246,9 @@ public class EurekaJerseyClientImpl implements EurekaJerseyClient {
                     TrustManager[] trustManagers = factory.getTrustManagers();
 
                     sslContext.init(null, trustManagers, null);
-                    SSLSocketFactory sslSocketFactory = new SSLSocketFactory(sslContext);
-                    sslSocketFactory.setHostnameVerifier(SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
+                    X509HostnameVerifier hostnameVerifier = SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER;
+                    SSLConnectionSocketFactory customSslSocketFactory = new SSLConnectionSocketFactory(sslContext, hostnameVerifier);
+                    SSLSocketFactory sslSocketFactory = new SSLSocketFactoryAdapter(customSslSocketFactory);
                     SchemeRegistry sslSchemeRegistry = new SchemeRegistry();
                     sslSchemeRegistry.register(new Scheme(PROTOCOL, HTTPS_PORT, sslSocketFactory));
 
@@ -257,6 +263,18 @@ public class EurekaJerseyClientImpl implements EurekaJerseyClient {
                         }
                     }
                 }
+            }
+
+            /**
+             * @see SchemeRegistryFactory#createDefault()
+             */
+            private MonitoredConnectionManager createDefaultSslCM() {
+                final SchemeRegistry registry = new SchemeRegistry();
+                registry.register(
+                        new Scheme("http", 80, PlainSocketFactory.getSocketFactory()));
+                registry.register(
+                        new Scheme("https", 443, new SSLSocketFactoryAdapter(SSLConnectionSocketFactory.getSocketFactory())));
+                return new MonitoredConnectionManager(clientName, registry);
             }
         }
     }

--- a/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/SSLSocketFactoryAdapter.java
+++ b/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/SSLSocketFactoryAdapter.java
@@ -1,0 +1,131 @@
+package com.netflix.discovery.shared.transport.jersey;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.security.cert.X509Certificate;
+
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+
+import org.apache.http.HttpHost;
+import org.apache.http.client.HttpClient;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.conn.ssl.SSLSocketFactory;
+import org.apache.http.conn.ssl.X509HostnameVerifier;
+import org.apache.http.protocol.HttpContext;
+
+/**
+ * Adapts a version 4.3+ {@link SSLConnectionSocketFactory} to a pre 4.3
+ * {@link SSLSocketFactory}. This allows {@link HttpClient}s built using the
+ * deprecated pre 4.3 APIs to use SSL improvements from 4.3, e.g. SNI.
+ *
+ * @author William Tran
+ *
+ */
+public class SSLSocketFactoryAdapter extends SSLSocketFactory {
+
+    private final SSLConnectionSocketFactory factory;
+
+    public SSLSocketFactoryAdapter(SSLConnectionSocketFactory factory) {
+        // super's dependencies are dummies, and will delegate all calls to the
+        // to the overridden methods
+        super(DummySSLSocketFactory.INSTANCE, DummyX509HostnameVerifier.INSTANCE);
+        this.factory = factory;
+    }
+
+    @Override
+    public Socket createSocket(final HttpContext context) throws IOException {
+        return factory.createSocket(context);
+    }
+
+    @Override
+    public Socket connectSocket(
+            final int connectTimeout,
+            final Socket socket,
+            final HttpHost host,
+            final InetSocketAddress remoteAddress,
+            final InetSocketAddress localAddress,
+            final HttpContext context) throws IOException {
+        return factory.connectSocket(connectTimeout, socket, host, remoteAddress, localAddress, context);
+    }
+
+    @Override
+    public Socket createLayeredSocket(
+            final Socket socket,
+            final String target,
+            final int port,
+            final HttpContext context) throws IOException {
+        return factory.createLayeredSocket(socket, target, port, context);
+    }
+
+    private static class DummySSLSocketFactory extends javax.net.ssl.SSLSocketFactory {
+        private static final DummySSLSocketFactory INSTANCE = new DummySSLSocketFactory();
+
+        @Override
+        public Socket createSocket(Socket s, String host, int port, boolean autoClose) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public String[] getDefaultCipherSuites() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public String[] getSupportedCipherSuites() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Socket createSocket(InetAddress host, int port) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Socket createSocket(String host, int port, InetAddress localHost, int localPort)
+                throws IOException, UnknownHostException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort)
+                throws IOException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class DummyX509HostnameVerifier implements X509HostnameVerifier {
+        private static final DummyX509HostnameVerifier INSTANCE = new DummyX509HostnameVerifier();
+
+        @Override
+        public boolean verify(String hostname, SSLSession session) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void verify(String host, SSLSocket ssl) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void verify(String host, X509Certificate cert) throws SSLException {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void verify(String host, String[] cns, String[] subjectAlts) throws SSLException {
+            throw new UnsupportedOperationException();
+        }
+
+    }
+
+}

